scalafix-rules/src/main/scala/scalafix/internal/rule/RemoveUnused.scala

```diff
@@ -29,27 +31,42 @@ class RemoveUnused
           "Run `scalac -Ywarn-unused:help` for more details. " +
           s"Obtained ${config.scalacOptions}")
     } else {
-      Configured.ok(this)
+      config.conf
+        .getOrElse("RemoveUnused")(this.config)
+        .map(new RemoveUnused(_))
     }
 
   override def fix(implicit doc: SemanticDoc): Patch = {
     val isUnusedTerm = mutable.Set.empty[Position]
     val isUnusedImport = mutable.Set.empty[Position]
 
-    doc.diagnostics.foreach {
-      case message if message.message == "Unused import" =>
-        isUnusedImport += message.position
-      case message if isUnusedPrivateDiagnostic(message) =>
-        isUnusedTerm += message.position
-      case _ =>
+    doc.diagnostics.foreach { diagnostic =>
+      if (config.imports && diagnostic.message == "Unused import") {
+        isUnusedImport += diagnostic.position
+      } else if (config.privates &&
+        diagnostic.message.startsWith("private") &&
+        diagnostic.message.endsWith("is never used")) {
+        isUnusedTerm += diagnostic.position
+      } else if (config.locals &&
+        diagnostic.message.startsWith("local") &&
+        diagnostic.message.endsWith("is never used")) {
+        isUnusedTerm += diagnostic.position
+      } else {
+        () // ignore
+      }
     }
 
-    doc.tree.collect {
-      case i: Importee if isUnusedImport(importPosition(i)) =>
-        Patch.removeImportee(i).atomic
-      case i: Defn if isUnusedTerm(i.pos) =>
-        defnTokensToRemove(i).map(Patch.removeTokens).asPatch.atomic
-    }.asPatch
+    if (isUnusedImport.isEmpty && isUnusedTerm.isEmpty) {
+      // Optimization: don't traverse if there are no diagnostics to act on.
```

Comment:
cc/ @ShaneDelmore making use of lazy trees here :) running `RemoveUnused` on the entire akka codebase (~220k loc) takes ~20 seconds on my laptop. That includes loading up all semanticdbs and producing a 7k loc diff.

Reply:
That is so cool to see 👍